src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java

```diff
@@ -715,6 +716,73 @@ public void visitForLoop(JCForLoop tree) {
         }
     }
 
+    @Override
+    public void visitForeachLoop(JCTree.JCEnhancedForLoop tree) {
+        bindingContext = new BasicBindingContext();
+        VarSymbol prevCurrentValue = currentValue;
+        try {
+            if (tree.getDeclarationKind() == EnhancedForLoopTree.DeclarationKind.PATTERN) {
+                /**
+                 * A statement of the form
+                 *
+                 * <pre>
+                 *     for (<pattern> : coll ) stmt ;
+                 * </pre>
+                 *
+                 * (where coll implements {@code Iterable<R>}) gets translated to
+                 *
+                 * <pre>{@code
+                 *     for (<type-of-coll-item> N$temp : coll) {
+                 *     switch (N$temp) {
+                 *         case <pattern>: stmt;
+                 *         case null: throw new MatchException();
+                 *     }
+                 * }</pre>
+                 *
+                 */
+                Type selectorType = tree.varOrRecordPattern.type;
+
+                currentValue = new VarSymbol(Flags.FINAL | Flags.SYNTHETIC,
+                        names.fromString("patt" + tree.pos + target.syntheticNameChar() + "temp"),
+                        selectorType,
+                        currentMethodSym);
+
+                JCStatement newForVariableDeclaration =
+                        make.at(tree.pos).VarDef(currentValue, null).setType(selectorType);
+
+                List<JCExpression> params = List.of(makeNull(), makeNull());
```

Comment:
The specification says the exception should be `new MatchException(new NullPointerException())`, but the nested NPE seems to be missing here.

FWIW, I think it is OK, for now, to create an empty NPE, but we may be able to do something better after PR 9746.

Reply:
Addressed!